PPO/PPO.py

```diff
@@ -67,3 +87,85 @@ def learn(self, num_learn):
       self.optimiser.step()
     
     self.model_old.load_state_dict(self.model.state_dict())
+
+class PPOPixel(PPOBase):
+  def __init__(self, config):
+    super(PPOPixel, self).__init__(config)
+    self.config = config
+  
+  def state_shaper(self, state):
+    state = np.array(state).transpose((2, 0, 1))
+    state = torch.FloatTensor(state)
+    state = state.unsqueeze(0)
+    state = state.float() / 256
+
+    return state
+  
+  def add_to_mem(self, state, action, reward, log_prob, done):
+    state = self.state_shaper(state)
+    self.mem.add(state, action, reward, log_prob, done)
+  
+  def act(self, x):
+    x = self.state_shaper(x).to(self.config.device)
+    return self.model_old.act(x)
+
+  def learn(self, num_learn, last_value, next_done):
+    # Calculate discounted rewards
+    bootstrap_length = self.config.update_every
+    discounted_returns = torch.zeros(bootstrap_length)
+    for t in reversed(range(bootstrap_length)):
+      # If first loop
+      if t == bootstrap_length - 1:
+        nextnonterminal = 1.0 - next_done
+        next_return = last_value
+      else:
+        nextnonterminal = 1.0 - self.mem.dones[t+1]
+        next_return = discounted_returns[t+1]
+      discounted_returns[t] = self.mem.rewards[t] + self.gamma * nextnonterminal * next_return
+
+    # normalise rewards
+    discounted_returns = torch.FloatTensor(discounted_returns).to(self.device)
+    discounted_returns = (discounted_returns - discounted_returns.mean()) / (discounted_returns.std() + 1e-8)
+    
+    # Fetch collected state, action and log_probs from memory & reshape for nn
+    prev_states = torch.stack(self.mem.states).reshape((-1,)+(4, 84, 84)).to(self.device).detach()
+    prev_actions = torch.stack(self.mem.actions).reshape(-1).to(self.device).detach()
+    prev_log_probs = torch.stack(self.mem.log_probs).reshape(-1).to(self.device).detach()
+
+    for i in range(num_learn):
+
+      # find ratios
+      actions, log_probs, values, entropy = self.model.act(prev_states, prev_actions)
+      ratio = torch.exp(log_probs - prev_log_probs.detach())
+      values = values.squeeze() * 0.5
+
+      # Stats
+      approx_kl = (prev_log_probs - log_probs).mean()
+
+      # calculate advantage & normalise
+      advantage = discounted_returns - values
+      advantage = (advantage - advantage.mean()) / (advantage.std() + 1e-8)
```

Comment:
`1e-8` - excellent stuff

Reply:
noice, we should definitely make this configurable swell